File: sdk/go/synapse/synapse.go
package synapse

import (
    "errors"
    "math"
)

const PhaseStep = 0.61803398875

type Feature struct { BaselineDelta float64; Quality float64 }
type Update struct { Vector []float64; Revision uint64; Confidence float64 }
type State struct { dimensions int; leak float64; inputGain float64; vector []float64; revision uint64 }

func New(dimensions int, leak, inputGain float64) (*State, error) {
    if dimensions <= 0 { return nil, errors.New("dimensions must be positive") }
    if math.IsNaN(leak) || math.IsInf(leak,0) || leak < 0 || leak >= 1 { return nil, errors.New("leak must be in [0,1)") }
    if math.IsNaN(inputGain) || math.IsInf(inputGain,0) || inputGain < 0 { return nil, errors.New("inputGain must be non-negative") }
    return &State{dimensions:dimensions, leak:leak, inputGain:inputGain, vector:make([]float64,dimensions)}, nil
}

func New12D() *State { state,_ := New(12,0.88,0.12); return state }

func (s *State) Update(features []Feature, confidence float64) (Update,error) {
    if math.IsNaN(confidence) || math.IsInf(confidence,0) || confidence < 0 || confidence > 1 { return Update{},errors.New("confidence must be in [0,1]") }
    for _,f := range features { if math.IsNaN(f.BaselineDelta) || math.IsInf(f.BaselineDelta,0) || math.IsNaN(f.Quality) || math.IsInf(f.Quality,0) || f.Quality < 0 || f.Quality > 1 { return Update{},errors.New("invalid feature") } }
    s.revision++
    if len(features)==0 { return Update{Vector:append([]float64(nil),s.vector...),Revision:s.revision,Confidence:0},nil }
    inputs:=make([]float64,len(features)); for i,f:=range features { inputs[i]=math.Tanh(f.BaselineDelta)*f.Quality }
    next:=make([]float64,s.dimensions); for i:=range next { source:=inputs[i%len(inputs)]; phase:=math.Sin(float64(i+1)*PhaseStep); value:=s.leak*s.vector[i]+s.inputGain*source*phase; next[i]=math.Max(-1,math.Min(1,value)) }
    s.vector=next; return Update{Vector:append([]float64(nil),next...),Revision:s.revision,Confidence:confidence},nil
}

File: src/cosmos_bio_cns/synapse.py
from __future__ import annotations

from dataclasses import dataclass
import math
from typing import Iterable, Sequence

PHASE_STEP = 0.61803398875
DEFAULT_DIMENSIONS = 12
DEFAULT_LEAK = 0.88
DEFAULT_INPUT_GAIN = 0.12


@dataclass(frozen=True)
class SynapticFeature:
    """Language-neutral feature used by the deterministic synaptic kernel."""

    baseline_delta: float
    quality: float

    def __post_init__(self) -> None:
        if not math.isfinite(self.baseline_delta):
            raise ValueError("baseline_delta must be finite")
        if not math.isfinite(self.quality) or not 0.0 <= self.quality <= 1.0:
            raise ValueError("quality must be finite and in [0,1]")


def _coerce_feature(feature: SynapticFeature | tuple[float, float]) -> SynapticFeature:
    if isinstance(feature, SynapticFeature):
        return feature
    baseline_delta, quality = feature
    return SynapticFeature(float(baseline_delta), float(quality))


def synaptic_step(
    previous_state: Sequence[float],
    features: Iterable[SynapticFeature | tuple[float, float]],
    *,
    leak: float = DEFAULT_LEAK,
    input_gain: float = DEFAULT_INPUT_GAIN,
) -> tuple[float, ...]:
    """Run one deterministic COSMOS synaptic state transition.

    This is an engineering interoperability primitive. It is not a medical,
    diagnostic, biological, or consciousness model.
    """

    if not previous_state:
        raise ValueError("previous_state must not be empty")
    if not math.isfinite(leak) or not 0.0 <= leak < 1.0:
        raise ValueError("leak must be finite and in [0,1)")
    if not math.isfinite(input_gain) or input_gain < 0.0:
        raise ValueError("input_gain must be finite and non-negative")

    state = tuple(float(value) for value in previous_state)
    if not all(math.isfinite(value) for value in state):
        raise ValueError("previous_state values must be finite")

    normalized = tuple(_coerce_feature(feature) for feature in features)
    if not normalized:
        return state

    inputs = tuple(math.tanh(feature.baseline_delta) * feature.quality for feature in normalized)
    next_state: list[float] = []
    for i, previous in enumerate(state):
        source = inputs[i % len(inputs)]
        phase = math.sin((i + 1) * PHASE_STEP)
        value = leak * previous + input_gain * source * phase
        next_state.append(max(-1.0, min(1.0, value)))
    return tuple(next_state)


def cosmos_12d_step(
    previous_state: Sequence[float],
    features: Iterable[SynapticFeature | tuple[float, float]],
) -> tuple[float, ...]:
    if len(previous_state) != DEFAULT_DIMENSIONS:
        raise ValueError("cosmos_12d_step requires exactly 12 state values")
    return synaptic_step(previous_state, features)
